syntax = "proto3";

package example1;

message TypeMessage {
  // Scalar value types.
  double double_field = 1;
  float float_field = 2;
  int32 int32_field = 3;
  int64 int64_field = 4;
  uint32 uint32_field = 5;
  uint64 uint64_field = 6;
  sint32 sint32_field = 7;
  sint64 sint64_field = 8;
  fixed32 fixed32_field = 9;
  fixed64 fixed64_field = 10;
  sfixed32 sfixed32_field = 11;
  sfixed64 sfixed64_field = 12;
  bool bool_field = 13;
  string string_field = 14;
  bytes bytes_field = 15;

  // Enumerations.
  enum Enum {
    ENUM_INVALID = 0;
    ENUM_ONE = 1;
    ENUM_TWO = 2;
  }
  Enum enum_field = 16;

  // Nested types.
  message NestedMessage {
    double double_field = 1;
    string string_field = 2;
  }
  NestedMessage msg_field = 17;

  // Optional fields.
  optional int32 optional_field = 18;
  oneof oneof_field_outer {
    int32 oneof_field = 19;
  }

  // Repeated fields.
  repeated int32 int32_repeated = 20;
  repeated NestedMessage msg_repeated = 21;

  // Map fields.
  map<int32, NestedMessage> map_field = 22;

  // Reserved fields.
  reserved 23;
  reserved "reserved_field";
}
